### PR TITLE
[dagster-airlift] static assets operator

### PR DESCRIPTION
**Pull request by dpeng817**
## Changelog

Insert changelog entry or "NOCHANGELOG" here.

- [ ] `NEW` _(added new feature or capability)_
- [ ] `BUGFIX` _(fixed a bug)_
- [ ] `DOCS` _(added or updated documentation)_

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25092?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#25092** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25092?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#25091** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25091?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25090** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25090?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25089** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25089?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25088** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25088?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25087** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25087?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/static_assets_operator.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+from airflow.utils.context import Context
+
+from dagster_airlift.in_airflow.base_proxy_operator import (
+    BaseProxyToDagsterOperator,
+    DefaultSessionProxyToDagsterOperator,
+)
+
+
+class BaseStaticAssetsOperator(BaseProxyToDagsterOperator):
```
Comment (dpeng817) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/static_assets_operator.py:11]: MaterializeDagsterAssetsOperator

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            )
+        yield from [path_str_to_node[get_asset_key_str(path)] for path in self.asset_key_paths]
+
+
+def get_asset_key_str(path: Sequence[str]) -> str:
+    return "/".join(path)
+
+
+class DefaultMaterializeAssetsOperator(BaseMaterializeAssetsOperator):
+    """The default static assets operator - which opens a blank session and expects the dagster URL to be set in the environment.
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:39]: change docstring to match class name

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+from dagster_airlift.in_airflow.base_asset_operator import BaseDagsterAssetsOperator
+
+
+class BaseMaterializeAssetsOperator(BaseDagsterAssetsOperator):
+    """An operator base class that proxies execution to a user-provided list of Dagster assets.
+    Will throw an error at runtime if not all assets can be found on the corresponding Dagster instance.
+    """
+
+    def __init__(self, asset_key_paths: Sequence[Union[str, Sequence[str]]], *args, **kwargs):
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:15]: document key path format

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        }
+        if not all(get_asset_key_str(path) in path_str_to_node for path in self.asset_key_paths):
+            raise ValueError(
+                f"Could not find all asset key paths {self.asset_key_paths} in the asset nodes. Found: {list(path_str_to_node.keys())}"
+            )
+        yield from [path_str_to_node[get_asset_key_str(path)] for path in self.asset_key_paths]
+
+
+def get_asset_key_str(path: Sequence[str]) -> str:
+    return "/".join(path)
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:35]: This isn't safe

AssetKey("//", "/") and AssetKey("/", "//") would cause a collision

File: examples/experimental/dagster-airlift/dagster_airlift_tests/integration_tests/materialize_assets_operator_test_project/dags/dag.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+dag = DAG(
+    "the_dag",
+    default_args=default_args,
+    schedule_interval=None,
+    is_paused_upon_creation=False,
+    start_date=datetime(2023, 1, 1),
+)
+the_task = DefaultMaterializeAssetsOperator(
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift_tests/integration_tests/materialize_assets_operator_test_project/dags/dag.py:27]: `Default` is a bit of an odd prefix for a operator we expect to instantiate directly

**Review by schrockn — CHANGES_REQUESTED** [review]
enough to do a spin

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            )
+        yield from [path_str_to_node[get_asset_key_hash(path)] for path in self.asset_key_paths]
+
+
+def get_asset_key_hash(path: Sequence[str]) -> int:
+    """Asset keys are represented by a sequence of strings. This function converts the sequence to a hashable object, which allows us to use it as a key in a dictionary."""
+    return hash(tuple(path))
+
+
+class BlankSessionAssetsOperator(BaseMaterializeAssetsOperator):
```
Comment (benpankow) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:44]: What's the purpose of this operator? Is it just for tests & if so can we move it to the test class?

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+from dagster_airlift.in_airflow.base_asset_operator import BaseDagsterAssetsOperator
+
+
+class BaseMaterializeAssetsOperator(BaseDagsterAssetsOperator):
+    """An operator base class that proxies execution to a user-provided list of Dagster assets.
+    Will throw an error at runtime if not all assets can be found on the corresponding Dagster instance.
+
+    Args:
+        asset_key_paths (Sequence[Union[str, Sequence[str]]]): A list of asset key paths to materialize.
+            Each asset key is expected to be a bare string (which is converted to an asset key path with a single element)
```
Comment (benpankow) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:16]: nit: this is still a bit hard to parse - would be good to use the same language our docs do (e.g. `component` instead of `element`) - & maybe link out to docs? https://docs.dagster.io/concepts/assets/software-defined-assets#multi-component-asset-keys

> Each path in the list can be a string, which is treated as an asset key path with a single component, or a list of strings representing a path with multiple components

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        }
+        if not all(get_asset_key_hash(path) in path_str_to_node for path in self.asset_key_paths):
+            raise ValueError(
+                f"Could not find all asset key paths {self.asset_key_paths} in the asset nodes. Found: {list(path_str_to_node.keys())}"
+            )
+        yield from [path_str_to_node[get_asset_key_hash(path)] for path in self.asset_key_paths]
+
+
+def get_asset_key_hash(path: Sequence[str]) -> int:
+    """Asset keys are represented by a sequence of strings. This function converts the sequence to a hashable object, which allows us to use it as a key in a dictionary."""
```
Comment (benpankow) [inline examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:40]: You can use `tuple(path)` as a dict key, since tuples are hashable, which might be slightly nicer (e.g. map keys are still recognizable as the asset key path they represent)

File: examples/experimental/dagster-airlift/dagster_airlift_tests/integration_tests/materialize_assets_operator_test_project/dags/dag.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    default_args=default_args,
+    schedule_interval=None,
+    is_paused_upon_creation=False,
+    start_date=datetime(2023, 1, 1),
+)
+the_task = BlankSessionAssetsOperator(
+    # Test both string syntax and list of strings syntax.
+    task_id="some_task",
+    dag=dag,
+    asset_key_paths=["some_asset", ["other_asset"]],
```
Comment (benpankow) [inline examples/experimental/dagster-airlift/dagster_airlift_tests/integration_tests/materialize_assets_operator_test_project/dags/dag.py:31]: should we test a multi-component key path too?

**Review by benpankow — APPROVED** [review]
few inline comments

File: examples/experimental/dagster-airlift/dagster_airlift_tests/integration_tests/materialize_assets_operator_test_project/dags/dag.py (hunk 9 above)
Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift_tests/integration_tests/materialize_assets_operator_test_project/dags/dag.py:31]: yea we probably should for completion's sake.

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py (hunk 8 above)
Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:40]: oh nice. Yea that conversion would be much nicer.

File: examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py (hunk 6 above)
Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift/in_airflow/materialize_assets_operator.py:44]: Ultimately yea it's just for tests I guess.

**Comment by dpeng817** [comment]
### Merge activity

* **Oct 10, 10:04 AM PDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/25092).
